### ISSUE TITLE
Tuple/ROW comparison not supported in Oracle and SQL Server

### ISSUE THREAD

**Issue by Fire-Dragons** [djrobstep/sqlakeyset#52]
The official explanation of Sqlalchemy is as follows: The composite IN construct is not supported by all backends, and is currently known to work on PostgreSQL, MySQL, and SQLite. Unsupported backends will raise a subclass of DBAPIError when such an expression is invoked.

**Comment by acarapetis** [comment]
Could you please describe the issue you're experiencing? I'm unsure what relevance the passage you've quoted has to sqlakeyset, since we do not make use of `in_` anywhere.

**Comment by Fire-Dragons** [comment]
> 您能否描述您遇到的问题？我不确定您所引用的段落与slackyset有什么关联，因为我们没有在`in_`任何地方使用。

> q = User.query().order_by(User.nickname, User._id)
> page1 = get_page(q, per_page=5)
> next_page = page1.paging.next
> page2 = get_page(q, per_page=5, page=next_page)

page2 raise an error:
sqlalchemy.exc.DatabaseError: (cx_Oracle.DatabaseError) ORA-01796: this operator cannot be used with lists
[SQL: SELECT user__id, user__create_time, user__update_time, user__operator, user__version, user__status, user_nickname, user_gender, user_birthday, user_first_name, user_last_name, user_head_id, user_widgets_json
FROM (SELECT "user"."_id" AS user__id, "user"."_create_time" AS user__create_time, "user"."_update_time" AS user__update_time, "user"."_operator" AS user__operator, "user"."_version" AS user__version, "user"."_status" AS user__status, "user".nickname AS user_nickname, "user".gender AS user_gender, "user".birthday AS user_birthday, "user".first_name AS user_first_name, "user".last_name AS user_last_name, "user".head_id AS user_head_id, "user".widgets_json AS user_widgets_json
FROM "user"
WHERE ("user".nickname, "user"."_id") > (:param_1, :param_2) ORDER BY "user".nickname ASC, "user"."_id" ASC)

The root cause is that sql server and oracle do not support the operation of tuples

My solution is:

> def where_condition_for_page(ordering_columns, place, dialect):
>     """Construct the SQL condition required to restrict a query to the desired
>     page.
> 
>     :param ordering_columns: The query's ordering columns
>     :type ordering_columns: list(:class:`.columns.OC`)
>     :param place: The starting position for the page
>     :type place: tuple
>     :param dialect: The SQL dialect in use
>     :returns: An SQLAlchemy expression suitable for use in ``.where()`` or
>         ``.filter()``.
>     """
>     if len(ordering_columns) != len(place):
>         raise InvalidPage(
>             "Page marker has different column count to query's order clause"
>         )
> 
>     zipped = zip(ordering_columns, place)
>     swapped = [c.pair_for_comparison(value, dialect) for c, value in zipped]
>     row, place_row = zip(*swapped)
> 
>     if len(row) == 1:
>         condition = row[0] > place_row[0]
>     elif dialect.name.lower() not in ('postgresql', 'mysql', 'sqlite'):
>         condition = and_(*[row[index]>place_row[index] for index in range(len(row)) ])
>     else:
>         condition = tuple_(*row) > tuple_(*place_row)
>     return condition

**Comment by acarapetis** [comment]
OK, so it looks like tuple comparison is not supported in Oracle. We probably won't be adding official support for Oracle, since it's difficult to set up automated testing; but we can probably fix this problem.

> `and_(*[row[index]>place_row[index] for index in range(len(row)) ])`

This is not the correct expression to achieve the equivalent of `ORDER BY` on multiple columns - for example, it would not say that `(2,1)>(1,2)` is true. The correct comparison for `A>B` is something like

```
(A[0] > B[0]) OR
(A[0] == B[0] AND A[1] > B[1]) OR
(A[0] == B[0] AND A[1] == B[1] AND A[2] > B[2]) OR ...
```

We should be able to build this dynamically for dialects that don't support tuple comparison. If you want to give it a try, feel free to open a PR; otherwise I'll get to it when I have some free time.

**Comment by acarapetis** [comment]
Fixed in #53.
